zboxcore/sdk/deleteworker.go

```diff
@@ -26,109 +26,145 @@ type DeleteRequest struct {
-		return
+		return err
 	}
 
 	ctx, cncl := context.WithTimeout(req.ctx, (time.Second * 30))
-	_ = zboxutil.HttpDo(ctx, cncl, httpreq, func(resp *http.Response, err error) error {
+	return zboxutil.HttpDo(ctx, cncl, httpreq, func(resp *http.Response, err error) error {
 		if err != nil {
 			l.Logger.Error("Delete : ", err)
 			return err
 		}
 		defer resp.Body.Close()
 		if resp.StatusCode == http.StatusOK {
-			req.consensus.Done()
-			deleteMutex.Lock()
-			req.deleteMask |= (1 << uint32(blobberIdx))
-			deleteMutex.Unlock()
-			l.Logger.Info(blobber.Baseurl, " "+req.remotefilepath, " deleted.")
-		} else if resp.StatusCode == http.StatusNoContent {
-			req.consensus.Done()
-			deleteMutex.Lock()
-			req.deleteMask |= (1 << uint32(blobberIdx))
-			deleteMutex.Unlock()
-			l.Logger.Info(blobber.Baseurl, " "+req.remotefilepath, " not available in blobber.")
-		} else {
-			resp_body, err := ioutil.ReadAll(resp.Body)
-			if err == nil {
-				l.Logger.Error(blobber.Baseurl, "Response: ", string(resp_body))
-			}
+			l.Logger.Info(blobber.Baseurl, " "+req.remoteFilePath, " deleted.")
+			return nil
+		}
+
+		if resp.StatusCode == http.StatusNoContent {
+			l.Logger.Info(blobber.Baseurl, " "+req.remoteFilePath, " not available in blobber.")
+			return nil
+		}
+
+		resp_body, err := ioutil.ReadAll(resp.Body)
+		if err == nil {
+			l.Logger.Error(blobber.Baseurl, "Response: ", string(resp_body))
 		}
-		return nil
+
+		return fmt.Errorf("delete: %s", resp.Status)
 	})
+
 }
 
 func (req *DeleteRequest) getObjectTreeFromBlobber(blobber *blockchain.StorageNode) (fileref.RefEntity, error) {
-	return getObjectTreeFromBlobber(req.ctx, req.allocationID, req.allocationTx, req.remotefilepath, blobber)
+	return getObjectTreeFromBlobber(req.ctx, req.allocationID, req.allocationTx, req.remoteFilePath, blobber)
+}
+
+func (req *DeleteRequest) deleteFileFromBlobber(b *blockchain.StorageNode) (fileref.RefEntity, error) {
+
+	refEntity, err := req.getObjectTreeFromBlobber(b)
+	if err != nil {
+		if errors.Is(err, constants.ErrNotFound) {
+			req.consensus.Done()
+			return nil, nil
+		}
+
+		return nil, err
+	}
+
+	err = req.deleteBlobberFile(b)
+	if err != nil {
+		return nil, err
+	}
+
+	req.consensus.Done()
+	return refEntity, nil
 }
 
 func (req *DeleteRequest) ProcessDelete() error {
 	num := len(req.blobbers)
+	numNotFound := 0
 	objectTreeRefs := make([]fileref.RefEntity, num)
-	var deleteMutex sync.Mutex
-	removedNum := 0
-	req.wg = &sync.WaitGroup{}
-	req.wg.Add(num)
+
+	wait := make(chan DeleteResult, num)
+
 	for i := 0; i < num; i++ {
 		go func(blobberIdx int) {
-			defer req.wg.Done()
-			refEntity, err := req.getObjectTreeFromBlobber(req.blobbers[blobberIdx])
+
+			fr, err := req.deleteFileFromBlobber(req.blobbers[blobberIdx])
 			if err == nil {
-				req.consensus.Done()
-				deleteMutex.Lock()
-				req.listMask |= (1 << uint32(blobberIdx))
-				objectTreeRefs[blobberIdx] = refEntity
-				deleteMutex.Unlock()
+				util.WithRecover(func() {
```

Comment:
The util.WithRecover is not needed. We can define a `wg := sync.WaitGroup{}` and call `defer wg.Done()` in each goroutine function, then below the for loop, we call the `wg.Wait()` to wait for all the goroutines finish, till then should we start to read the `wait` channel, and we would not have panic issue.

Reply:
fixed